docs/rfc/rfc-027-p2p-message-bandwidth-report.md

```diff
@@ -0,0 +1,137 @@
+Google Cloud charges between [$.05 to $.12 per gigabyte of egress traffic][gcloud-pricing], and ingress is free. Hetzner [charges 1€ per TB used over the 10-20TB base bandwidth per month][hetzner-pricing], which will be easily hit if multiple terabytes are sent and received per day. Using the values collected from the validator on Osmosis, a single node on Google cloud may cost $18 to $44 a day running on Google cloud. On Hetzner, the estimated 18TB a month of both sending and receiving may cost between 0 and 10 Euro a month per node.
+
+## Discussion
+
+### Overview of Major Bandwidth Usage
+
+To determine which components of Tendermint were consuming the most bandwidth, I gathered prometheus metrics from the [Blockpane][blockpane] validator running on the Osmosis network for several hours.
+The data reveal that three message types account for 98% of the total bandwidth consumed. These message types are as follows:
+
+1. [consensus.BlockPart][block-part-message]
+2. [mempool.Txs][mempool-txs-message]
+3. [consensus.Vote][vote-message]
+
+
+The image below of p2p data collected from the Blockpane validator illustrates the bandwidth consumption of these three message types in both data sending and receiving.
+
+![](images/total-message-send-rate.png)
+![](images/total-message-recv-rate.png)
+
+### Investigation of Message Usage
+
+This section discusses the usage of each of the three highest consumption messages.
+
+#### BlockPart Transmission
+
+Sending `BlockPart` messages consumes the most bandwidth out of all p2p messages types as observed in the Blockpane Osmosis validator.
+In the almost 3 hour observation, the validator sent about 20 gigabytes of `BlockPart` messages.
+
+A block is proposed each round of Tendermint consensus. The paper does not define a specific way that
+the block is to be transmitted, just that all participants will receive it via a gossip network.
+
+The Go implementation of Tendermint transmits the block in 'parts'. It serializes
+the block to wire-format proto and splits this byte representation into a set of
+4 kilobyte arrays and sends these arrays to its peers, each in a separate message.
+
+The logic for sending `BlockPart` messages resides in the code for the [consensus.Reactor][gossip-data-routine].
+The consensus reactor starts a new `gossipDataRoutine` for each peer it connects to.
+This routine repeatedly picks a part of the block that Tendermint believes the peer does not know about yet and gossips it to the peer.
+The set of `BlockParts` that Tendermint considers its peer as having is only updated in one of four ways:
+
+1. [We receive a block part from the peer][block-part-receive].
+2. [We send][block-part-send-1] [the peer a block part][block-part-send-2].
+3. Our peer tells us about the parts they have block [via `NewValidBlock` messages][new-valid-block-message-send]. This message is only sent when the validator has a quorum of prevotes or precommits for a block.
```

Comment:
```suggestion
3. Our peer tells us about the block parts they have [via `NewValidBlock` messages][new-valid-block-message-send]. This message is only sent when the validator has a quorum of prevotes or precommits for a block.
```

Reply:
> This message is only sent when the validator has

Do you mean the peer? (is it always a validator?)

Reply:
Good catch, fixed.